### ISSUE TITLE
[FEATURE REQUEST] - Merge multiple databases and once with the option to have redundant databases deleted.

### ISSUE THREAD

**Issue by rtevans01** [keepassxreboot/keepassxc#5024]
I use Syncthing to sync my database with the all devices I use. Unfortunately, I often run into the issue of getting conflicting databases. I think this is solely due to the nature of Syncthing since it's not centralized and can get messy. 

I'm requesting the existing merging feature in KeePassXC be expanded to open multiple conflicting databases at once, merge them, and then automatically delete them afterwards. This would make maintenance easier for me. I don't know if this is a common issue for most other users though.

**Comment by aisbergde** [comment]
@rtevans01 
How keepassxc should know which databases belongs together?

My main issues also are related to synchronizing databases on different devices using cloud services.

Maybe the request could be more general:
- define databases which belongs together (maybe using some syntax in filenames)
- open all databases which belongs together in different tabs, requesting the password only one time
- allow merging between all opened databases, not requesting a password again
- add a context menu to tabs of opened database:
    - close tab
    - delete database
    - ...

**Comment by rtevans01** [comment]
Yes I think a syntax should be used. My conflicted databases are all in the same directory path as my non-conflicted database, so I don't think it would be very difficult to search for them. Just start out with the name of the non-conflicted database and look for any other database that has the same name plus "sync-conflict-2020...". 

I think opening separate tabs would be a waste of time. I'd rather have it taken care of in one step but to each his own.

**Comment by boernsen1** [comment]
After reading a lot of related issues, I think there is quiet some confusion in the acual issues and responsibilities related to cloud syncing in combination with KeePassXC, but I think this issue hits it best, so I am adding some details to it. After analysing it for quiet some time now, let me use this comment to document it in detail and hopefully also motivate for a new feature of KeePassXC as I think it will affect many users out there.

I think using KeePassXC on multiple devices accessing a single remote kdbx file works seemlessly as long as direct file system access to the actual and only kdbx file is used (such as an SMB share)! Changes done to the file while still open in KeePassXC can be automatically merged if configured. If the file is not accessible anymore on the other hand, KeePassXC will notify you. It therefore only works properly though in local networks or anywhere where connectivity to the original kdbx file is stable, so that you can use a file system based access to the file. We use this method in a team already since a year and I recommend everyone who is able to use this method to use it as it's easy to setup, very reliable and has great usability as changes are merged immediately. Thank you for this awesome feature!

However, at the point where KeePassXC is supposed to be used with cloud sync (like Dropbox, Syncthing, etc.), there is one major issue making it very annoying to use as described already by @rtevans01. It is actually not a bug, but a major usability issue due to the nature of how cloud sync works and how tasks/responsibilities are distributed across different softwares (such as KeePass and a sync software in this case).

The main issue are conflicts happening whenever two devices do not have a continuous connection. See this example with devices A and B:
- A and B both have a connection, their sync'ed kdbx files are up-to-date
- A and B both loose connection
- Both make any update or change and save their file locally (independent which entry they change or whether they add or remove something)
- Now the sync provider tries to sync the files on both devices
- Let's say A comes first for whatever reason and succeeds syncing
- B tries the same, but cannot sync to the actual kdbx file anymore since it has changed already. This is called a sync conflict, so the sync tool usually prevents data loss by creating a conflict file next to the actual file containing the local changes, which then is synced to both devices, A and B again. The actual kdbx (with A's changes) is then synced to B also causing B's KeePassXC to detect the file changes (if still opened) and asking for a reload or even reloading it automatically if configured. However, the actual kdbx now only contains A's changes and from B's point of view B's local changes seem to have disappeared suddenly. [#usability_issue1] They are still in the conflict file of course.
- Now, A and B have two versions of the kdbx, an original (containing the changes from A) and a conflict file (containing the changes from B)
- To resolve this conflict, firstly, either A or B needs to actually notice the existence of a conflict file [#usability_issue2] and then manually merge it into the actual kdbx, save the actual kdbx and delete the conflict kdbx. [#usability_issue3]
- The merged kdbx will then be synced to the other device again and both are in sync again.

The general situation is not unusual as in practice the connection between the device and a cloud server is not always reliable. Think of mobile phones or home-driven cloud servers for example. With a LAN-only Syncthing setup, this even is every day practice.

Note that I tagged some positions with #usability_issueX. I think the major usability issues are:
- [#usability_issue1] The automatic reload is correct if configured, but very confusing as the local changes just disappear. Not having the automatic reload makes distributed changes very inconvenient. If KeePassXC was closed during the sync conflict, B probably wouldn't even notice the revert except B manually checks for conflict files every time opening the database.
- [#usability_issue2] The conflict itself can be overseen easily since you have to manually check for conflict files
- [#usability_issue3] The manual merge/delete is inconvenient if you need to do it often and error prone if there is more than a single conflict file as you need to manually compare file names.

Note that the file name patterns for conflict files are well documented for the popular sync providers:
- https://docs.syncthing.net/users/syncing.html?highlight=conflict#conflict-handling
- `<filename>.sync-conflict-<date>-<time>-<modifiedBy>.<ext>`
- https://help.dropbox.com/files-folders/share/conflicted-copy
- `“Conflicted copy”, the editor’s username, and the save date will be added to the file name.`
- https://docs.nextcloud.com/desktop/2.5/conflicts.html or https://doc.owncloud.org/desktop/2.5/conflicts.html
- `mydata (conflicted copy 2018-04-10 093612).txt containing “local contents”`

I therefore would like to encourage to add new features to KeePassXC that increase the usability for this very regular usecase of combining KeePassXC with cloud sync. I am not too familiar with best practices and no-gos in terms of security, but here are some ideas I have:
- automatic detection of conflict files (eventually by a manually configurable file name pattern to put this responsibility to the user)
- automatic merge/move-to-bin of conflict files
- alternatively (if the actions are not supposed to be done automatically for whatever reason) at least automatic notifications/dialogs about the existance of conflict files, the merge and/or the move-to-bin action
- at very least a good documentation about the situation with cloud sync

I hope this clarifies and motivates for some solution of the usability issue.

That was a long run. Thanks for reading down to here! :)

I was tempted to open a new issue for that, but I think there is quiet some confusion already due to many related tasks and some of the aspects described are already somewhere else, so I don't wanna mess it up more. These are some related tasks which might be worth considering: #90, #2937, #4192, #5072, #5024.
What would be the best practice with this @droidmonkey?

**Comment by droidmonkey** [comment]
Can you TLDR the above

**Comment by aisbergde** [comment]
@boernsen1 
You described this very clear. A "TLDR" makes not so much sense, because then some important content and context will be missing (IMHO). Thank you for this comment.

And that's why my solution with Keepass was (and is)
- to use a "master" file which is synced between devices. 
- I don't change the master directly, but I change "working copies" on different devices.
- I explicitly "merge" (in keepass words: synchronize) my working copy with the master when the connection is good and there is a good change to get the master updated in the cloud.

That's why I still hope to get bi-directional merge.

And, to go back to my comment above: KeepassX should try to use the same password as for the current opened file. This is how Keepass works: I will not be asked to enter the password again for each file I want to synchronize. I can just "synchronize" and that's it.

**Comment by boernsen1** [comment]
I'll try to sum up my indeed very long comment.

### Current situation
- Sharing a kdbx with direct file access / single file (like SMB) works great and very convenient due to the automatic and robust merge feature!
- Sharing a kdbx with any cloud sync (Syncthing, Dropbox, Nextcloud, etc) in theory works, but is very inconvenient and prone to oversee things (thus almost unusable) due to regularly arising sync conflict files, which have to be merged manually in an inconvenient process in KeePassXC. For details why those conflict files arise so regularly, see my detailed example above.

### Inconvenient due to
- manually checking for conflict files / sudden disappearing of changes just made
- manually selecting conflict files for merge
- manually entering the (identical) password of the conflict file
- manually deleting already merged conflict files by comparing complicated file names

### Desired
- convenient handling of these conflict files (see above for patterns), preferably automatic, but at least semi-automatic using dialogs and notifications

**Comment by droidmonkey** [comment]
Great thank you

**Comment by boernsen-development** [comment]
Hey @droidmonkey, is there any news about that?

At least there should be a comment about it at: https://keepassxc.org/docs/#faq-cloudsync
It still suggests, there is nothing unusual to expect, but if you're not a technician, I claim one would not know where to find missing (conflicting) passwords.

**Just to make it clear: cloud synchronisation currently only works flawlessly if only one device is writing and all others are reading only. Otherwise conflicts are very likely, especially with deferred sync tools like Syncthing.**

**Comment by droidmonkey** [comment]
No there hasn't been any effort put forth on this one. Personally, I use cloud sync all the time and write to my database on mobile device and laptop, never had a conflict yet. This doesn't appear to be a common problem for our user base, this is the only discussion about sync conflicts we ever had.

**Comment by ekasprzak** [comment]
Can't agree with that - whenever I try (every few months in last few years) to switch back from Bitwarden (because of no auto-type function there) I ALWAYS get conflicts in matter of max few days, where most of the time it happens much quicker.
Fully agree with @boernsen1 that suggesting that it works great is just wrong.
Quick search on google (and even ddg) for keepassxc conflicts on sync brings thousands of results with a lot of people having those problems.
So, I disagree completely with "This doesn't appear to be a common problem for our user base"
It's a big blocker for me for switch back to Keepassxc (I did the switch few years ago exactly because of constant sync issues even when I'm trying hard to follow general advice - only one writing to db device at the time)

**Comment by droidmonkey** [comment]
You might be confusing the request to have a direct cloud sync solution with this request for conflict resolution. Also the original keepass does not have this conflict resolution either so it's not like we are alone here.

**Comment by aisbergde** [comment]
bad synchronization support is the main issue in KeepassXC.

The normal Keepass doesn't have this issue, because it can sync in both directions. And it doesn't require to reenter the password again and again, when both databases have the same password.

Keepass for Android also can sync in in both directions, but it needs to be online, and this is sometimes not the case.

Currently, you must be careful if you want to change the database on different devices and there is no connection to the cloud in the meantime. In fact, it only works if you make sure that changes to the database end up in the cloud immediately.

As soon as you change the database on any device that is not online, there are often problems. And it's when you're on the road and you change the database on a mobile device and the changes don't immediately end up in the cloud that the problems start.

Therefore, one solution would be to use a master database that is not changed directly, but with which device-specific databases are synchronized. For me would be OK, if this is not done in one step, but in 2 steps:
- merge from the other database
- merge into the other database

But then at least it should be recognized that both databases have the same password, so you must enter it only once: in the database you are working with.

**Comment by tomwj** [comment]
I've had this issue intermittently for the last couple of years using Syncthing.

This thread has been useful for some workarounds e.g. making sure at least one device is always on and running with Syncthing to avoid deferred sync, and confirms my suspicions about what was causing it.

The ability to merge databases automatically would be a handy feature for me, given it seems like the heavy lifting has already been done by the Merge database feature. However if it's not a common issue, it could be resolved by mentioning it in the docs when choosing a sync provider. https://keepassxc.org/docs/#faq-cloudsync

Really nice piece of software that keeps getting better. Thanks for all the effort 🙂

**Comment by jovandeginste** [comment]
I'll just chime in with my experience.

I used to use Dropbox for syncing, but switched to syncthing when Dropbox enforced their nr. of devices on Linux. My workflow is probably what triggers this; I have:

- a Linux home server which is almost always on
- an Android phone (always on and logged on) and several tablets (not usually logged on)
- a Linux desktop (work)
- a Linux laptop (work)
- a Windows laptop (gaming)

... which all share that one syncthing volume where my `kdbx` file is on, and where I use either KeepassXC (Linux/Windows) or Keepass2Android. I almost never edit the `kdbx` file on Android, I often (daily or more) edit it on the desktop or laptops.

Since I don't like waiting too much, I usually suspend/hibernate the desktop and laptops when I'm done using them. This is probably what triggers the merge problem.

My hunch:

Is it possible that KeepassXC saves an update to disk when it comes out of suspend mode (because time has passed), before syncthing had a chance of syncing?

**Comment by droidmonkey** [comment]
Make sure your syncthing is actually syncing your database when it is changed. Are you using always save after every change?

**Comment by jovandeginste** [comment]
I do

**Comment by jovandeginste** [comment]
I can share my settings, if you wish

**Comment by droidmonkey** [comment]
Most likely syncthing daemon not actually syncing in a timely manner. We don't hold a lock on files, as soon as it is written we release the file.

**Comment by jovandeginste** [comment]
I suddenly had a hunch and tested it. I realized the sync-issue started manifesting itself after I configured KeepassXC Secret Service Integration.

The options are rather limited, here are my settings:

```ini
[FdoSecrets]
Enabled=true
NoConfirmDeleteItem=false
ShowNotification=false
```

I think one of the programs using the Secret Service (probably Evolution) is updating a secret right after coming back from hibernation, which is then saved in KeepassXC, before syncthing had a change of syncing the remote changes back.

Not sure who to "blame" here...

On a tangential note, I don't really like this Secret Service thing too much, since it was a serious PITA to kill the default in Gnome and have KeepassXC reliably fulfil that role...

**Comment by droidmonkey** [comment]
Ok, neither of those problems are our fault.

**Comment by boernsen-development** [comment]
Just another attempt from my side to make this issue and possible solutions more obvious.

### Issue:
* If using read-write cloud sync for a single database on multiple devices, which are likely to loose connection, conflict files are likely to appear without the user noticing it. This is very likely for decentral tools such as Syncthing in combination with multiple devices.
* These conflict files are likely to be overseen (thus passwords seem to suddenly disappear) and merging them into the actual database is very inconvenient.

### Workarounds:
* Use a dedicated database for each device, which is edited only on that device (read-write) and merge changes from other devices manually while treating other devices' databases as read-only. (all databases are synced between the devices)
* Manually check for conflict files regularly and merge them manually. This requires to enter the same password again.

### Some suggestions for development (collected from the discussions above):
* Mention the issue at https://keepassxc.org/docs/#faq-cloudsync
* Add automatic detection for conflict files. This should be fairly easy as they all contain the original filename and "conflict" (case-insensitive) somewhere in the filename. See my comment above for documentations of various sync providers.
* Add dialog noticing about existing conflict files or mention them on the unlock screen.
* Add feature to also open/unlock all related conflict databases when unlocking a database
* Add feature to re-use the password of an unlocked database when merging a locked database.
* Add feature to merge multiple databases at once.
* Add feature to optionally delete the merged databases (i.e. move them to bin).

I think it's not necessary to implement all of them. Maybe just the simplest ones would be enough to overcome the usability drawbacks of finding the conflict databases, merging them one by one and entering the same password again and again.
I hope this way we'll find some solutions or usability features that are simple enough to be picked for development one day.

Please let me know if you have more suggestions and I'll add them. Alternatively, the OP could also add such a list to the description if he wants to.

**Comment by erikvanoosten** [comment]
IMHO the *simplest thing* that can be implemented to overcome the usability problem are these 2 small features:

* Allow *multiple select* when selecting the file(s) for Merge Database, and then merge the selected databases one by one.
* When opening the to-be-merged database, *only ask for a password* if it can not be opened with the credentials of the current database. (Variations: 1. do show the password dialogue, but pre-fill the password, 2. display an 'are-you-sure' prompt.)

These features can be implemented independently, and they bring usability value independently as well.

**Comment by droidmonkey** [comment]
Good ideas!

**Comment by tobiasBora** [comment]
@erikvanoosten I would say that it is also important to **automatically** (this is important since most of the time people won't even realize there is a conflict) start the merge process whenever keepassxc sees that a new conflicting file has been created, potentially with a message:
```
Warning: keepassxc detected a conflict because the database was modified on another device. Do you want to automatically merge the two databases (X.kdbx and X_conflict.kbdx) to get rid of the conflict?
[Yes/No/Yes but do not remove the conflicted file/Yes but keep a backup of both files]
```



By default we could automatically search for conflict files using the [syntax used by famous cloud providers](https://github.com/keepassxreboot/keepassxc/issues/5024#issuecomment-760301354), and provide to the user a function to add a regex to find conflicting files (I guess we can search for files in the current directory only, it is certainly quicker and safer). This regex could certainly be saved in the database, just see the below security concern in that case. 

In term of security implication, I can only imagine one scenario (really special I guess) where storing the regex in the database could be a problem:
- if an adversary A is able to share a password database D with a party B
- if B stores another private database P in the same folder
- If B chooses exactly the same password for P and D

then A could learn the database P by adding the regex `.*` in D (or if B named its database with the word `conflict` in it… which is quite unlikely). Then P would be merged into D since they have the same password, and A would learn the passwords of B.

Of course, this attack is quite unlikely to happen in a real life since B has no reasons to use the same password for both databases and put them in the same folder… (with a regular cloud provider both files would have been shared to B which is already an issue). But anyway this "attack" can be mitigated by asking for a confirmation if the database uses a custom regex, while printing a warning if a database is created with a name containing the word `conflict`.

**Comment by dolfandringa** [comment]
> IMHO the _simplest thing_ that can be implemented to overcome the usability problem are these 2 small features:
> 
>     * Allow _multiple select_ when selecting the file(s) for Merge Database, and then merge the selected databases one by one.
> 
>     * When opening the to-be-merged database, _only ask for a password_ if it can not be opened with the credentials of the current database. (Variations: 1. do show the password dialogue, but pre-fill the password, 2. display an 'are-you-sure' prompt.)
> 
> 
> These features can be implemented independently, and they bring usability value independently as well.

I agree with these solutions. They solve most. For me the last step would be to automate the merge and record the paths of the databases to auto-merge. That would allow me to open my local copy, and merge with paths in the FS when opening/saving the database, if those files are available.

**Comment by Prophet-Phoenix** [comment]
Long time user of KeePassXC here. Love it, thank you! ❤️
Also using syncthing for quite some time and i also have syncconflicts when making fast changes to the kdbx file that aren't picked up by syncthing fast enough and not synced to all devices. And to be clear i think it is just syncthings limitations and it's not keepassxc at fault.

What i am missing in merging kdbx files is the option to not import deleted entrys again.
So if i deleted an entry "github.com" on computer A and earlier changed an entry "google.com" on computer B (which was not synced to computer A because of syncthing), the deleted entry "github.com" would get merged back in A, when merging the conflicting file B, rigth? Coming from B, "github.com" is a new entry. All other merges would be made correctly judging by date of modification.

Maybe show new entrys with older modification or creation date and ask which ones to import?

**Comment by droidmonkey** [comment]
> the deleted entry "github.com" would get merged back in A, when merging the conflicting file B, rigth?

It shouldn't. When you delete an entry, a record of that deletion is stored in the database file (the UUID of the entry). This is used to "re-delete" the entry if it is ever seen again on a merge. This is actually the root cause of #6477

**Comment by Prophet-Phoenix** [comment]
I tested again and turns out that my original kdbx file does not have a recycle bin, as the new ones have. I created a new database and merged with my original one and now it behaves like you say. So thanks!

Does the merge functionality rely on the recycle bin folder? It does feel a little odd that all my internet password history is in the recycle bin forever now. But i can live with that probably or delete them for good, as long as i dont have conflicting database files lying around.

**Comment by droidmonkey** [comment]
Recycle Bin doesn't mean the entry was deleted, it was just moved to that group. The deleted UUID only gets written when it is actually deleted. When you empty your recycle bin this will occur.

**Comment by serpro69** [comment]
I'm not sure I completely understand the problem with sync of multiple databases.
But maybe what I'm doing myself isn't actually doing what I expect it to?

Say I have two devices: A and B.

On each device, I have it's own .kdbx file. I do sync only in one direction, and then merge that with the "origin". For example:

- A has a.kdbx open and B has b.kdbx open locally
- A makes changes to a.kdbx and sends it to B
- B does merge of b.kdbx with a.kdbx
- B sends b.kdbx to A
- A does merge of a.kdbx with b.kdbx
- both A and B now have their own updated .kdbx files, which have identical contents

Am I doing something wrong? (apart from the fact that it does have a manual step to merge things, which I don't really much mind, at least at this point)

**Comment by droidmonkey** [comment]
@serpro69 that is certainly a way to do it that would avoid A and B conflicting with each other on writes (ie, they both make a change at nearly the same time before a sync can occur). However, if that is not possible because you are the only one making changes, then you are actually wasting your time. Just use one database and sync (send the file) between A and B.

**Comment by serpro69** [comment]
Thanks for the reply @droidmonkey !
No, I don't have situations where changes to both might occur at the same time. But the problem for me is that A and B aren't always running at the same time and hence changes might not always be synced right away. I use syncthing for file synchronization, so I need to actually power on both to sync when I need to. Until I can set-up a third device which would always run and through which others could sync, I don't see how I could always sync between A and B directly w/o accidentally loosing something.
Let's say I make a change on B, which isn't synced, and then I make a change on A which also isn't synced (and this I do have sometimes). When I do sync, wouldn't that result in a conflict because both files have diverged from a common point? I'm not sure how syncthing would resolve that, but I don't think that it would be too happy about it :D So my assumption is that in this case I would still need to fix this manually with a merge via keepassxc.

**Comment by droidmonkey** [comment]
You could setup a raspberry pi as a third party. Syncthing should create a conflict file, it does keep a version history anyway. You can always test with a dummy database.

**Comment by ustulation** [comment]
I too use syncthing to sync multiple devices (android, linux, mac etc) and was looking for a master-local sync. Every time I search for an optimal way I stumble upon a few open (like this current issue) and closed but not-actually-solved (like [this](https://github.com/keepassxreboot/keepassxc/issues/90) and I share[ this sentiment](https://github.com/keepassxreboot/keepassxc/issues/90#issuecomment-547215354) from there). My current strategy (I won't be surprised if it's sub-optimal) is as follows:

Goal:
1. Goal: The sync software (like syncthing) should be pretty dumb. In-fact it could just simply overwrite the files according to timestamps and there should still be no data loss, even if the keepass-* (KXC, K2A etc) app was offline.
2. It should work in a p2p scenario, I don't want my db in cloud and I don't want centralisation.

With that in mind, here's what I do:

Say I need to share a db with Linux (M0), Android (M1) and Mac (M2) -> db-0. I want to share another db-1 with only M0 and M1 but not M2 (because say M2 is company machine with MDMs and IT monitoring tools (aka corporate spywares) and I don't want personal stuff there).

1. Start with M0. Create KXC db there - call it `local-M0`. Password=P. Create two groups and call them db-0 and db-1.
2. Enable Keeshare for each and choose to bi-directionally sync with your main current db (local-M0).
3. When you choose keeshare it'll ask you for a path to dump the keeshared-db. So those paths are `/path/db-0` and `/path/db-1`
4. `db-0` path is shared via syncthing with `M1` and `M2`. `db-1` is shared only with `M1`.
5. Also when you choose keeshare, you will need to provide passwords for db-0 and db-1. Let's say for db-0 we use a new password `P0` and for db-1 we reuse the same `P` as our `local-M0`.
6. Use similar setup on `M2` (mac) except that it only has one group and thus one keeshare which is `db-0`. The password for the main db there (`local-M2`) and the keeshare `/some/path/db-0` is `P0`.
7. Now KXC on `M0` (linux) and `M2` (Mac) will monitor changes made to `db-0` and `db-1` (only `M0`) and do a bi-directional sync with their respective locals (`local-M0`, `local-M2`).

For `M1` (android):
1. Use K2A to point to the network sync'd `db-0` and `db-1` (the master db's that travel and get overwritten during network sync). The passwords are as you guessed `P0` and `P` respectively.
2. K2A has a feature called work on internal cache only. Activate that for both `db-0` and `db-1` (on `M1`). This forces K2A to copy these db's to some internal location and only read-write from there. It doesn't matter if (network sync'd copies of) `db-0` and `db-1` on `M1` have changed. So it has it's own `local-M1-db-0` and `local-M1-db-1` that you don't actually see (unless you explore the Android filesystem).

Now imagine a conflict:
1. No network
2. K2A modifies entry `x -> x-M1` in `local-M1-db-0` (remember it's set to work on internal cache only mode). Press `sync database` option from menu so that it actually writes changes to its `db-0`. Press `work of internal cache only` immediately again to return back to local mode. Then the app dies. No one knows about this modification yet.
3. KXC on `M0` modifies `x -> x-M0`. Due to the nature of KXC, it's immediately written to keeshare `db-0` too in addition to `local-M0`, but since the network is down, no one knows. KXC is closed/killed
4. KXC on `M2` creates a new entry `y`. Both `local-M2` and keeshare `db-0` is modified there, but no one else knows. KXC is closed/killed.
5. Bring the network back up.
6. You will see a few conflicts. But since we are imagining dumb-sync, just imagine your blindly delete all conflict files.
7. Let's say the winner was `x -> x-M0` which is now the `db-0` on all machines.
8. Bring all KXCs and K2A up. `M0` sees there's no change, does nothing.
9. `M2` KXC immediately realises the local and master are different, so it syncs again. Now it's `db-0` contains both `x -> x-M0` and `y` changes. This is picked up by syncthing which syncs again.
10. `M0` realises the change in keeshare and syncs everything again and now both `M0` and `M2` share the same world view.
11. Because things are not automatic in K2A unlike KXC, you will need to press `sync dbs` again from menu which will sync the local cache with the `db-0` (containing `M0's` and `M2's` changes now) and thus take in the new stuff and give it it's own `x -> x-M1` change. Depending on whether `x -> x-M0`'s timestamp was more recent than `x -> x-M1`'s timestamp or not, one becomes the latest and the other goes into the version history.
12. Now since `db-0` has changed again, it's resync'd and `M0` and `M2` pick up the changes.

So net net there's a lot of churn and eventually all get the same world view. Completely decentralize, p2p and no manually management of any sync-conflict - just imagine you instruct syncthing to not even produce a conflict file, just overwrite with whatever it thinks is latest etc, and things will still work and no data loss!

---

To me this is very powerful and so far has worked with zero data losses. In-fact [I made a youtube 4 part series](https://www.youtube.com/watch?v=1_RyD6vsYJo&list=PLk1_IP-XmS3hU39DH26PaLxxtp1AgD8Pg) about this simulation to show how it deals with various cases (Note: I'm not a YouTuber, so it's a boring/technical video).

Drawbacks:
1. K2A's work on internal cache only needs to be set each time after you press the "sync with main db"  option, otherwise it doesn't go back to that mode and starts directly interacting with the master db, which obviously you don't want because that can be overwritten when K2A is offline due to network-sync and all un-sync'd changes exclusively from K2A will be lost! I've created an issue about this in K2A's github repo, but it's still open.
2. While keeshare helps simulate this master-local in KXC, it has drawbacks like it only supports flat hierarchy AFAIK. There is an open issue here somewhere about it. So if you keeshare a group, say `db-0`, which has a sub-tree of folders and files, it gets flattened at the destination (or something like that) so don't do it. It's only if your are happy to have flat single level entries in each of your keeshares (`db-0`, `db-1` etc). I don't like it, but can live with it for now. I just prefix the entry name with what my folder structure would look like. So instead of `db-0 -> bank -> transferwise` I'll do something like `db-0 -> bank-transferwise` so when I type `bank` I've all the relevant entries listed. Would be awesome if KXC supported/preserved trees within keeshared groups while bi-directional syncs.

**Comment by boernsen-development** [comment]
I could imagine this all ends up in a decision by the KeePassXC devs on how and where to set the boundaries of KeePassXC's responsibility to automate user desires.

One could argue that all this "can" be done already, but just with poor software support and low UX, so that most users won't take the step. Whether this is relevant or not then depends on how many users are actually affected I guess.

To offer a more automated solution with minimal effort, I could imagine to use a separate detection/file-watch tool for conflict files, which then could trigger a notification or even a script event. KeePassXC could then increase automation and therefore increase UX by including the following features into GUI and/or CLI (as mentioned by @erikvanoosten already):

- multiselection of databases to be merged
- auto-reuse of password when merging other databases
- ask for deletion of merged databases 

The latter is not even necessary for the CLI as the external script (which has been triggered by a file-watcher and then triggers the KeePassXC CLI) could take care of it.
[Update: If I think about it, for the CLI, I think my proposal would NOT help as the password of a database opened in the GUI could not be reused by the CLI resulting in another password prompt except when using keyfiles or other password less methods. Although, one could actually use the GUI KeePassXC's Auto-Type for this if the DB's master password itself is stored in the DB if it is open in the GUI 😆]

This way, we would have a more automated solution with minimal effort at least for technical persons, who are able to set this up.

If we wanted to include more people into such a solution, I think we would need more UI support for the `detection` and `deletion` steps besides the proposed solution for the `merging` step. Generally I would categorize conflict file handling into three steps:

1. `detection`
2. `merging`
3. `deletion`

**Comment by codethief** [comment]
Another Syncthing user here, who has been struggling with this issue for a while. 

Just to throw it out there, since it hasn't been mentioned yet: If, under the hood, the KeePass database were implemented as a [CRDT](https://en.wikipedia.org/wiki/Conflict-free_replicated_data_type), it should be relatively straight-forward to merge several almost identical, yet diverging databases (of the type produced by e.g. Syncthing in the case of conflicts). Conflicts would no longer arise when two devices touch different parts of the database but only ever ensue when they modify the _same_ database entry independently from each other. In this case, KeePass could point out the conflict to the user while still allowing the user to modify the rest of the database. (In the same way as version control tools like jujutsu and pijul [have made conflits](https://jj-vcs.github.io/jj/latest/conflicts/) [part of their core data model](https://pijul.org/manual/conflicts.html), while git forces you to resolve conflicts right there and then before you can proceed). This would allow building a tool to fully automate the merging.

While I know almost nothing about the kdbx format, I'm sure this wouldn't be a trivial change, though.

**Comment by bickford** [comment]
> You might be confusing the request to have a direct cloud sync solution with this request for conflict resolution. Also the original keepass does not have this conflict resolution either so it's not like we are alone here.

Hi guys, love KeePassXC and the only issue I have with it when moving from KeePass is the problem described in this thread, I've read through it all and was surprised that nobody has mentioned that the original KeePass (v2) actually does have a robust solution to this problem.

Way back in the early days of Dropbox and cloud sync popularization, KeePass users were facing the same issue of conflicts between a database synched on multuple devices. Their solution was the following, which has worked flawlessly ever since:

1. Create a local copy of your database, separate but identical to the cloud copy in your shared folder.
2. Set your local copy to be opened by your KeePass2 (never the cloud copy).
3. In KP2 assign a trigger when this specific database is opened, to sync (merge) it with the copy in the cloud provider shared folder.
4. In KP2 assign a trigger when this specific database is closed, to sync (merge) it again with the copy in the cloud provider shared folder.

With this method you always work on your local copy only, which is then merged within KP (independent of the cloud provider) at opening and closing. Any changes made on the cloud copy by another device in the interim will also be synced. I have not had a conflict using this method for many years since.

I've read that there are dev concerns on other threads about implementing triggers, and even a CVE for KeePass (CVE-2023-24055) but I agree with the KP devs response that writing software with the assumption that the computer using the software could be hacked and preventing security issues in that scenario is impossible. There is absolutely no reason to presume your software itself is safe on a hacked system, the KeePassXC executable could just as relistically be overwritten and replaced by a malicious version as the config file, that does whatever a malicious actor wants it to - so any and all of your countermeasures would be rendered moot.

However, we can avoid the entire trigger discussion as the four-step solution above need not use a widely configurable user-defined trigger system, and could rather be an tightly-defined addition to the current options. For example, in Application Settings under 'File Management', something under 'Alternative saving method' could have "Work on a temporary database copy ONLY, then synchronize with real database file".

Is this way off the roadmap of your goals or is it a helpful suggestion?

**Comment by droidmonkey** [comment]
That is a really bad solution and adds complexity for no reason.

I have used keepassxc with onedrive (directly) for 10 years without EVER having a sync issue. This is with keepass2android and now keepassdx as well.

The problem is with the cloud sync software. Some vendors do it way better than others. Some are better on different platforms. 

We also offer alternate saves that can help prevent poorly behaved sync software from messing things up.

https://keepassxc.org/docs/KeePassXC_UserGuide#_advanced_save_options

**Comment by bickford** [comment]
I'm aware of KXC's alternate save options, which are no doubt very useful in some scenarios. But none of them keep a local cache copy and sync it back with the primary DB file - which goes a long way to eliminate issues of the database file being modified in multiple places at once in any cloud sync scenario. It is also one of the methods employed by the software you mention you've been using on Android - Keepass2android. They call it 'work on internal cache copy'.

The solution worked very well at the last business I worked for, team of five having a single cloud-shared Keepass2 file open and modified simultaneously daily and conflict-free for six years. This _is_ a more complex scenario than a single person using OneDrive, but it certainly doesn't add complexity for no reason - the reason is resolving the multiple trouble tickets over this issue, as others have mentioned in thread.

I expect you'll see this issue continue to rear its head, as SyncThing is an increasingly popular tool - but certainly not the only cloud-sync tool that people have conflict issues with when used to synchronize database files such as Keepass. In fact, even Microsoft _recommended not using OneDrive_ to share Access Database files (and presumably extended this advice to other types of database files) _due to the risk of corruption when database files are edited in multiple places_. https://techcommunity.microsoft.com/discussions/access/limitations-of-using-a-split-access-database-on-onedrive/2130037/replies/2130389

There is no way for a cloud sync solution - any of them - to determine 'which file wins' when files are modified in multiple places simultaneously; they just create a conflict file, flag a warning notification, and let the user figure it out later. The only way to avoid such issues is for the database software to handle such scenarios, so I disagree with this being laid at the feet of the cloud sync software. But i understand your position - dev time is finite and it's not something you value adding - at least you now have the information about how KeePass handled it (which wasn't in thread yet). Peace & I hope you have a great weekend.

**Comment by tobiasBora** [comment]
> I have used keepassxc with onedrive (directly) for 10 years without EVER having a sync issue

For my case I was having lot's of issues with synthing for a simple reason: the sync only happens when the two devices are connected at the same time (peer to peer synchronisation with no central server), but in practice it is very often the case that you edit stuff one one device, turn it off, open the other device (no sync happens because first device is off) and do an edit. This would always create a conflict, with the creation of a new conflicting file next to the database, require manual action to resolve it. So I agree with @bickford this is a true issue in some cases, as attested by the popularity of this bug.

@bickford : But I'm not sure to agree with your approach: this seems quite complicated (you need to maintain a cache…), and seems not very robust: say that device A and B modify the same file (both of them are offline, so keepassxc will still sync the cache with the local shared folder of each device), say that later when they reconnect syncthing keeps file A (from device A) and creates a conflicted file with the version of B. Then say B disconnects for a long time. Later, when A re-opens keepassxc, it will see that cache and shared file are equal, so no merge will occur, hence the conflicted file will be ignored.

For me cache only make sense when the sync just erases conflicting files… but most cloud providers will keep the conflict file.

So I don't understand: why don't we simply check if a conflicted file is present  when openning the database (configurable with regex by the user with sane defaults dealing with most known cloud providers), in which case we prompt a message to the user "the database XXX has another conflicting version myfile_conflict_20250201, do you want to merge them and remove the conflicted file? Yes/Always/No". Seems like super simple and quite nobust…

**Comment by droidmonkey** [comment]
> but in practice it is very often the case that you edit stuff one one device, turn it off, open the other device (no sync happens because first device is off) and do an edit. This would always create a conflict

This is true for any file then, this isnt a keepassxc problem. You set yourself up for failure. Introduce a third, always on, sync node. As it stands now, you are not syncing.

> why don't we simply check if a conflicted file is present when openning the database

What constitutes a conflict file? Every service does it differently. Do we clean up the mess, too? Do we worry about files older than the most recent? I am not interested in solving these problems when we have a merge feature that you can invoke at any time if you need it.

**Comment by tobiasBora** [comment]
> This is true for any file then, this isnt a keepassxc problem.

Sure, for other files one would need manual resolution, but keepassxc is clever enough to be able to automatically solve conflicts, why not exploit it?

> You set yourself up for failure. Introduce a third, always on, sync node. As it stands now, you are not syncing.

Well, not everyone can buy/maintain/configure a server, and some of us don't want to use proprietary solutions with data stored who knows where. And even with proprietary solutions, you can still end-up changing passwords without having access to internet (e.g. right now I'm in a place with intermittent internet, travellers also can't always have internet in foreign countries, countries in development also struggle with good connectivity…) If you don't have these issues, lucky you.

> What constitutes a conflict file? Every service does it differently

All services I know (nextcloud, syncthing, dropbox, google drive, …) create a new copy of the file with the conflicting content, and name it like `youroriginalfilename[Conflict].extension`, so a case insensitive regular expression like `${originalfilenamewithoutextension}.*conflict.*` should be able to detect most conflicts, and we can always let the user customize the regexp if they use a different tool.

> Do we clean up the mess, too? 

Sure, after a succesful merge no need to keep conflicting files. For extra sanity we can ask confirmation to the user, but this seems unnecessary to me.

> we have a merge feature that you can invoke at any time if you need it.

Well unless you spend all your time checking for conflicts you don't even know when you need it. The whole point of this issue is to automatically call the merge feature when needed to improve UX and avoid the burden of manually checking for conflicts + removing the database after the merge + type the password twice.

The fact that this feature is among the top 20 most asked features (based on reactions) indicates that this is something people need.

**Comment by thethinker990** [comment]
+1

> Sure, after a succesful merge no need to keep conflicting files. For extra sanity we can ask confirmation to the user, but this seems unnecessary to me.

In KeePass2 you could simply define some Trigger which solved all your problems. Only deleting the file would not be possible, but one could create a CRON/Windows Task Scheduler which deletes all files with a certain RegEx older than 2 weeks for example. 

For the problem being: I simply want to merge/sync (on open and save) with a primary database.

I was thinking mabye the KeeShare feature could lead to the goal. I set that up, shared root with a primary database but ended up with all files in the root folder. There was no mirroring/merge/sync, all groups forgotten. 

Also, if I create in db1 in groupA an entry1, and in db2 in group2 an entry2 they end up in each others root folder, which is VERY inconvenient. At least, when I edit in db1 entry2, the entry stays where it is, but only as long as I do not move it around in db2. Then it will end up again in root.

Even if I would end up creating a KeeShare for each group, I would ending up with 40+ files and 40+ passwords. All of them not properly secured, no option for challenge-response nor a keyfile for those KeeShare databases.

I also often have syncing problems with my cloud. The main problem is, the cloud does not know (thank god) what's happening inside my database, thus creating a 2nd copy of the file instead of overwriting it (thank god).

Conveniently I used the trigger system in KeePass2 to merge/sync my changes on open (if the primary database was updated/changed) and on save (to merge/sync my changes to the primary database).
If the authentication was exactly the same, the merge happened automagically without further authentication.

In KeePassXC the only option one has is, if one is having a sync-problem and a 2nd copy is created (since there is no option to have a synced primary database!), to manually merge the two databases.
